packages/core/echo/echo-db/src/proxy-db/mutable-schema-registry.ts

```diff
@@ -19,17 +21,167 @@ import { createStoredSchema } from '@dxos/live-object';
 import { log } from '@dxos/log';
 
 import { type EchoDatabase } from './database';
-import { getObjectCore } from '../echo-handler';
+import { getObjectCore, type ReactiveEchoObject } from '../echo-handler';
 import { Filter } from '../query';
 
 export type SchemaSubscriptionCallback = (schema: MutableSchema[]) => void;
 
-export interface SchemaResolver {
+export interface SchemaRegistry {
+  query(query: SchemaRegistryQuery): SchemaRegistryLiveQuery<SchemaRecord>;
+
+  /**
+   * Registers the provided schema.
+   *
+   * @returns Schema records after registration.
+   *
+   * The behavior of this method depends on the state of the database. The general
+   * principle is that the schema will be upserted into the space. If an equivalent
+   * schema with the same name and version already exists, the method does nothing.
+   * If a different schema with the same name and version exists, the method throws
+   * an error. If no schema with the same name and version exists, a new schema will
+   * be inserted based on semantic versioning rules.
+   */
+  register(input: RegisterSchemaInput[]): Promise<SchemaRecord[]>;
+
+  /**
+   * @deprecated Use `query()`.
+   */
   getSchema(typename: string): MutableSchema | undefined;
-  query(): Promise<MutableSchema[]>;
+
+  /**
+   * @deprecated Use `query()`.
+   */
   subscribe(cb: SchemaSubscriptionCallback): UnsubscribeCallback;
 }
 
+export type SchemaRegistryQuery = {
+  /**
+   * Filter by schema ID.
+   * Schema id is a DXN with `echo` or `type` kind.
+   */
+  id?: string[];
+
+  /**
+   * One or more typenames to filter by.
+   */
+  typename?: string[];
+
+  /**
+   * [Semver Range](https://docs.npmjs.com/cli/v6/using-npm/semver#ranges) for the schema version.
+   */
+  version?: string;
+};
+
+export interface SchemaRegistryLiveQuery<T> {
```

Comment:
SchemaRegistrySubscription?

Reply:
the type is reused for one-shot queries too

Reply:
the idea behind this (and also data queries in ECHO) is that you first create this LiveQuery object -- it sort-of acts like a prepared query in SQL databases, and then you either run it one-shot, or subscribe to it.

Reply:
I'll rename it PreparedQuery